shared/src/current_block.rs

```diff
@@ -0,0 +1,73 @@
+use anyhow::{anyhow, Context, Result};
+use futures::{stream::FusedStream, Stream};
+use primitive_types::H256;
+use std::time::Duration;
+use web3::{
+    types::{BlockId, BlockNumber},
+    Transport, Web3,
+};
+
+pub type Block = web3::types::Block<H256>;
+
+const POLL_INTERVAL: Duration = Duration::from_secs(1);
+
+/// A stream that yields the current block whenever it changes.
+pub fn block_stream(web3: Web3<impl Transport>) -> impl Stream<Item = Block> + FusedStream {
+    // We do not use a block filter because filters have been unreliable in the past, do not work
+    // with load balancing over multiple nodes and would still need to be polled unless we used web
+    // sockets.
+    futures::stream::unfold(H256::zero(), move |previous| {
+        let web3 = web3.clone();
+        async move {
+            loop {
+                tokio::time::delay_for(POLL_INTERVAL).await;
+                let block = match current_block(&web3).await {
+                    Ok(block) => block,
+                    Err(err) => {
+                        tracing::warn!("failed to get current block: {:?}", err);
+                        continue;
+                    }
+                };
+                let hash = match block.hash {
+                    Some(hash) => hash,
+                    None => {
+                        tracing::warn!("missing hash");
+                        continue;
+                    }
+                };
+                if hash == previous {
+                    continue;
+                }
+                return Some((block, hash));
```

Comment:
I don't think this is handling the case well when fetching the current_block taskes longer than the time between blocks (which theoretically can be less than 1s). From the documentation I would assume the stream to yield all block changes without any gaps (and a consumer might rely on the hashes returned to e.g. populate a database. 

I'm not sure we have to provide a gapless experience (by inspecting block.parent and comparing it with a lit of previous hashes) but should maybe clarify that this is a "latest block stream" which might have gaps.

Reply:
Agree, will document.